File: meet-up-parser/src/meet-up-parser.js
const createMeetUps = (input) => {
  return { meetUps: input.input.map(parseMeetUp) };
};

const parseMeetUp = (rawMeetUp) => {
  return parseEdition(rawMeetUp) +
    rawMeetUp.name +
    parseDates(rawMeetUp) +
    parseMeetUpLocation(rawMeetUp.location.map((location) => {
      return parseCityState(location) + location.country;
    }));
};

function parseEdition(rawMeetUp) {
  return rawMeetUp.edition ? rawMeetUp.edition + ' ' : '';
}

function parseDates(rawMeetUp) {
  return ' · ' +
    (rawMeetUp.startDate ? rawMeetUp.startDate : '') +
    (rawMeetUp.endDate ? ' / ' + rawMeetUp.endDate : '');
}

function parseMeetUpLocation(location) {
  if (location.length > 1) {
    if (location[0].split(', ')[1] === location[1].split(', ')[1]) {
      return (location[0] + location[1].replace(' · ', ' | ')).replace(', ' + location[0].split(', ')[1], '');
    } else {
      return location[0] + location[1].replace(' · ', ' | ');
    }
  } else {
    return location;
  }
}

function parseCityState(location) {
  return ' · ' +
    (location.city ? location.city + cityStateSeparator(location, 'city') : '') +
    (location.state ? location.state + cityStateSeparator(location, 'state') : '');
}

function cityStateSeparator(location, type) {
  if(location.city && location.state && type === 'city') return ', ';
  else if(location.city && location.state && type === 'state') return '. ';
  else if(!location.city && location.state && type === 'city') return '';
  else if(!location.city && location.state && type === 'state') return ', ';
  else return ', ';
}

exports.build = createMeetUps;

File: meet-up-java/src/main/java/com/joantolos/meetup/MeetUpParser.java
package com.joantolos.meetup;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class MeetUpParser {

    public String parseToJson(String inputJson) throws IOException {
        return new ObjectMapper().writeValueAsString(new Contract.Output(
            new ObjectMapper().readValue(inputJson, Contract.Input.class).input()
                .stream()
                .map(meetUp -> {
                    List<Contract.Location> locations = meetUp.location();
                    List<String> parsedLocations = new ArrayList<>();
                    for (int i = 0; i < locations.size(); i++) {
                        Contract.Location location = locations.get(i);
                        boolean isNextCountryTheSame = i + 1 < locations.size() && locations.get(i + 1) != null && locations.get(i + 1).country().equals(location.country());
                        parsedLocations.add(
                            (location.city() != null ? location.city() : "") + (location.state() == null && location.city() != null ? (!isNextCountryTheSame ? ", " : "") : "") +
                            ((location.state() != null && location.city() != null ? ", " : "") + (location.state() != null ? location.state() + (location.city() != null ? ". " : ", ") : "")) +
                            (!isNextCountryTheSame ? location.country() : "")
                        );
                    }
                    return (meetUp.edition() != null ? meetUp.edition() + " " : "") +
                        meetUp.name() + " · " +
                        meetUp.startDate() +
                        (meetUp.endDate() != null ? " / " + meetUp.endDate() + " · " : " · ") +
                        String.join(" | ", parsedLocations);
                }).collect(Collectors.toList())
        ));
    }
}

interface Contract {
    record Input(List<MeetUp> input) { }
    record MeetUp(String edition, String name, String startDate, String endDate, List<Location> location) { }
    record Location(String city, String state, String country) { }
    @JsonInclude(JsonInclude.Include.NON_NULL)
    record Output(List<String> meetUps) { }
}

